Sources/Penny/BotStateManager.swift

```diff
@@ -2,64 +2,123 @@ import DiscordBM
 import Foundation
 import Logging
 
-/*
+/**
  When we update Penny, AWS waits a few minutes before taking down the old Penny instance to
  make sure the new instance is healthy.
  This makes it so there is a short period where there are 2 Penny bots that will respond to
  Discord Gateway events.
  This actor's job is to prevent that. Only the new bot should respond to events.
+
+ This will:
+   * Disallow Gateway-event handling from the beginning.
+   * When `initialize()` is called, this will send a "please shutdown" message.
+   * When the old Penny instance receive that message.
+     * The old instance will cache save all needed cached stuff into a S3 bucket.
+     * Then it will send a "did shutdown" message.
+   * The new instance will catch the "did shutdown" message.
+     * The new instance will get the stuff from the S3 bucket.
+     * Then it will start allowing Gateway-event handlings.
+   * If the old instance is too slow to make the process happen, the process is aborted and
+     the new instance will start handling events without waiting more for the old instance.
  */
 actor BotStateManager {
     
     var logger = Logger(label: "BotStateManager")
-    var canRespond = true
+    var canRespond = false
     let id = Int(Date().timeIntervalSince1970)
-    
-    let signal = "Hello the other Pennys 👋 you can retire now :)"
+
+    var cachesService: any CachesService {
+        ServiceFactory.makeCachesService()
+    }
+
     var disableDuration = Duration.seconds(3 * 60)
-    
+
+    var isCachePopulated = false
+
     static private(set) var shared = BotStateManager()
     
     private init() { }
     
-    func initialize() {
+    func initialize() async {
         self.logger[metadataKey: "id"] = "\(self.id)"
-        Task { await sendSignal() }
+        Task { await send(.shutdown) }
+        cancelIfCachePopulationTakesTooLong()
     }
-    
+
+    func cancelIfCachePopulationTakesTooLong() {
+        Task {
+            try await Task.sleep(for: .seconds(15))
+            if !isCachePopulated {
+                canRespond = true
+                logger.error("No CacheStorage-population signal was done in-time")
+            }
+        }
```

Comment:
Traditionally one does this by keeping the `Task` handy, calling `cancel()` when populating is done, and having the task body be just the timeout case. This approach works but unnecessarily binds the task to the actor and can in theory suffer from a TOCTOU race (though it's quite extremely unlikely even for a race condition).